src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -652,6 +652,189 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+                {
+                    op2             = gtNewIconNode(0x80000000);
+                    simdBaseType    = TYP_INT;
+                    simdBaseJitType = CORINFO_TYPE_INT;
+                    vectorCreateOp1 = gtNewLconNode(0xFFFFFFE2FFFFFFE1);
+
+                    if (simdSize == 16)
+                    {
+                        vectorCreateOp2 = gtNewLconNode(0xFFFFFFE4FFFFFFE3);
+                    }
+                    break;
+                }
+
+                case TYP_DOUBLE:
+                {
+                    op2             = gtNewLconNode(0x8000000000000000);
+                    simdBaseType    = TYP_LONG;
+                    simdBaseJitType = CORINFO_TYPE_LONG;
+                    vectorCreateOp1 = gtNewLconNode(0xFFFFFFFFFFFFFFC1);
+
+                    if (simdSize == 16)
+                    {
+                        vectorCreateOp2 = gtNewLconNode(0xFFFFFFFFFFFFFFC2);
+                    }
+                    break;
+                }
+
+                default:
+                {
+                    unreached();
+                }
+            }
+
+            if (simdSize == 16)
+            {
+                op3 = gtNewSimdHWIntrinsicNode(simdType, vectorCreateOp1, vectorCreateOp2, NI_Vector128_Create,
+                                               vectorCreateType, simdSize);
+            }
+            else
+            {
+                op3 =
+                    gtNewSimdHWIntrinsicNode(simdType, vectorCreateOp1, NI_Vector64_Create, vectorCreateType, simdSize);
+            }
+
+            op1 = gtNewSimdHWIntrinsicNode(simdType, op1, op2, NI_AdvSimd_And, simdBaseJitType, simdSize,
+                                           /* isSimdAsHWIntrinsic */ false);
+
+            op1 = gtNewSimdHWIntrinsicNode(simdType, op1, op3, NI_AdvSimd_ShiftLogical, simdBaseJitType, simdSize,
+                                           /* isSimdAsHWIntrinsic */ false);
+
+            if (varTypeIsByte(simdBaseType) && (simdSize == 16))
+            {
+                CORINFO_CLASS_HANDLE clsHnd = gtGetStructHandleForSIMD(simdType, simdBaseJitType);
+
+                op1 = impCloneExpr(op1, &op2, clsHnd, (unsigned)CHECK_SPILL_ALL,
+                                   nullptr DEBUGARG("Clone op1 for vector extractmostsignificantbits"));
+
+                op1 = gtNewSimdHWIntrinsicNode(TYP_SIMD8, op1, NI_Vector128_GetLower, simdBaseJitType, simdSize,
+                                               /* isSimdAsHWIntrinsic */ false);
+                op1 = gtNewSimdHWIntrinsicNode(TYP_SIMD8, op1, NI_AdvSimd_Arm64_AddAcross, simdBaseJitType, 8,
+                                               /* isSimdAsHWIntrinsic */ false);
+                op1 = gtNewSimdHWIntrinsicNode(simdBaseType, op1, NI_Vector64_ToScalar, simdBaseJitType, 8,
+                                               /* isSimdAsHWIntrinsic */ false);
+                op1 = gtNewCastNode(TYP_INT, retNode, /* isUnsigned */ true, simdBaseType);
```

Comment:
Am I seeing things here or `retNode` is `nullptr` at this point?

Also, why `/* isUnsigned */ true`?

Reply:
Yes, fixed.

Its unsigned because zero extensions are cheaper and this does not need (or want) sign extension.